File: go/0036.valid-sudoku/solution.go
// Created by Olgierd Palasz at 2026/06/22 20:38
// leetgo: dev
// https://leetcode.com/problems/valid-sudoku/

package main

import (
	"bufio"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

func hashCheck(array []byte) bool {
	hash := make(map[byte]bool)

	for _, val := range array {
		if hash[val] && val != 46 { // byte('.') = 46
			return false
		} else {
			hash[val] = true
		}
	}
	return true
}

func isValidSudoku(board [][]byte) bool {
	// rows
	for _, row := range board {
		if !hashCheck(row) {
			return false
		}
	}

	// columns
	for i := 0; i < 9; i++ {
		array := make([]byte, 9)

		for j := 0; j < 9; j++ {
			array[j] = board[j][i]
		}

		if !hashCheck(array) {
			return false
		}
	}

	a := 0
	b := 0

	// 3x3 squares
	for k := 0; k < 9; k++ {
		array := make([]byte, 0, 9)

		for i := 0; i < 3; i++ {
			for j := 0; j < 3; j++ {
				array = append(array, board[a+i][b+j])
			}
		}

		if a == 6 {
			a = 0
			b += 3
		} else {
			a += 3
		}

		if !hashCheck(array) {
			return false
		}
	}

	return true
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	board := Deserialize[[][]byte](ReadLine(stdin))
	ans := isValidSudoku(board)

	fmt.Println("\noutput:", Serialize(ans))
}

File: rust/src/0036.valid-sudoku/solution.rs
// Created by Olgierd Palasz at 2026/07/17 17:39
// leetgo: dev
// https://leetcode.com/problems/valid-sudoku/

use anyhow::Result;
use leetgo_rs::*;

struct Solution;

// @lc code=begin
use std::collections::HashSet;
impl Solution {
    pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {
        let mut rows: Vec<HashSet<char>> = vec![HashSet::with_capacity(9); 9];
        let mut cols: Vec<HashSet<char>> = vec![HashSet::with_capacity(9); 9];
        let mut boxes: Vec<HashSet<char>> = vec![HashSet::with_capacity(9); 9];

        for i in 0..9 {
            for j in 0..9 {
                let val = board[i][j];

                if val == '.' {
                    continue;
                }

                let box_indx: usize = (i / 3) * 3 + (j / 3);

                if rows[i].contains(&val)
                    || cols[j].contains(&val)
                    || boxes[box_indx].contains(&val)
                {
                    return false;
                }

                rows[i].insert(val);
                cols[j].insert(val);
                boxes[box_indx].insert(val);
            }
        }

        true
    }
}

// @lc code=end

fn main() -> Result<()> {
    let board: Vec<Vec<char>> = deserialize(&read_line()?)?;
    let ans: bool = Solution::is_valid_sudoku(board);

    println!("\noutput: {}", serialize(ans)?);
    Ok(())
}
